### ISSUE TITLE
Attempt to read property "data" on null in user_node_load() (line 3041 of ../user.module)

### ISSUE THREAD

**Issue by alanmels** [backdrop/backdrop-issues#6375]
## Description of the bug

A Backdrop website with an imported DB (from another Backdrop site) with a simple Views page listing nodes started to flood the log with two warnings:

```
Warning: Attempt to read property "data" on null in user_node_load() (line 3041 of /var/www/html/docroot/core/modules/user/user.module).
Warning: Undefined array key 243 in user_node_load() (line 3041 of /var/www/html/docroot/core/modules/user/user.module).
```

## Steps To Reproduce

1. Login as any user other than user 1 and create bunch of nodes.
2. Delete that user directly from the database, for example, using phpMyAdmin.
3. Create any Views page that lists the affected nodes and see the logs flooded with the reported warnings.

## Additional information

Ideally all users who created nodes should exist, however in practice there can be all kinds of scenarios when it's not always true. In our case, two websites are not related at all, so we never sought to properly sync the users as the second website needed only large number of nodes created on the first. We resolved the problem by massively attributing nodes with nonexistent authors to Anonymous user, however I thought to report this so the community decides if it's a but at all and if it should be addressed to save other users from similar problems.

The offending function looks like below and it always assumes $node->uid exists:

```php
/**
 * Implements hook_node_load().
 */
function user_node_load($nodes, $types) {
  // Build an array of all uids for node authors, keyed by nid.
  $uids = array();
  foreach ($nodes as $nid => $node) {
    $uids[$nid] = $node->uid;
  }

  // Fetch name, picture, and data for these users.
  $user_fields = db_query("SELECT uid, name, picture, data FROM {users} WHERE uid IN (:uids)", array(':uids' => $uids))->fetchAllAssoc('uid');

  // Add these values back into the node objects.
  foreach ($uids as $nid => $uid) {
    $nodes[$nid]->name = $user_fields[$uid]->name;
    $nodes[$nid]->picture = $user_fields[$uid]->picture;
    $nodes[$nid]->data = $user_fields[$uid]->data;
  }
}
```

Changing it to could be one of possible solutions:

```php
/**
 * Implements hook_node_load().
 */
function user_node_load($nodes, $types) {
  // Build an array of all uids for node authors, keyed by nid.
  $uids = array();
  foreach ($nodes as $nid => $node) {
    if (null != user_load($node->uid)) {
      $uids[$nid] = $node->uid;
    }
  }

  // Fetch name, picture, and data for these users.
  $user_fields = db_query("SELECT uid, name, picture, data FROM {users} WHERE uid IN (:uids)", array(':uids' => $uids))->fetchAllAssoc('uid');

  // Add these values back into the node objects.
  foreach ($uids as $nid => $uid) {
    $nodes[$nid]->name = $user_fields[$uid]->name;
    $nodes[$nid]->picture = $user_fields[$uid]->picture;
    $nodes[$nid]->data = $user_fields[$uid]->data;
  }
}
```

But I'm not sure if it's a bug at all, so please share your insights and let's close it if it's not worth changing the core for this.

**Comment by indigoxela** [comment]
> But I'm not sure if it's a bug at all, so please share your insights and let's close it if it's not worth changing the core for this.

Good question. I'm not sure, whether handling that gracefully here is a good idea.

That a UID exists in the node object, but the user dataset doesn't is a failure on its own. This normally shouldn't happen. My concern is that if we catch it at this point, we'd obfuscate a data inconsistency - it might even get harder to discover the base problem.

**Comment by kiamlaluno** [comment]
Instead of logging four warnings for each non-existing account the actual code is producing (a warning for the non-existing array key, and three warnings for reading properties from `null`) I would rather log a single error.

```php
function user_node_load($nodes, $types) {
  // Build an array of all uids for node authors, keyed by nid.
  $uids = array();
  foreach ($nodes as $nid => $node) {
    $uids[$nid] = $node->uid;
  }

  // Fetch name, picture, and data for these users.
  $user_fields = db_query("SELECT uid, name, picture, data FROM {users} WHERE uid IN (:uids)", array(':uids' => $uids))->fetchAllAssoc('uid');

  // Add these values back into the node objects.
  foreach ($uids as $nid => $uid) {
    if (isset($user_fields[$uid])) {
      $nodes[$nid]->name = $user_fields[$uid]->name;
      $nodes[$nid]->picture = $user_fields[$uid]->picture;
      $nodes[$nid]->data = $user_fields[$uid]->data;
    }
    else {
      watchdog('user', 'Attempted to access a non-existing user account (%uid) while loading a node (%nid).', array('%uid' => $uid, '%nid' => $nid));
    }
  }
}
```

Backdrop core already logs an error when an account does not exist, in [`user_cancel()`](https://docs.backdropcms.org/api/backdrop/core%21modules%21user%21user.module/function/user_cancel/1).

```php
if (!$account) {
  backdrop_set_message(t('The user account %id does not exist.', array('%id' => $uid)), 'error');
  watchdog('user', 'Attempted to cancel non-existing user account: %id.', array('%id' => $uid), WATCHDOG_ERROR);
  return;
}
```

The log message would make clearer what is happening, and it is just a single log message instead of four warnings.

**Comment by argiepiano** [comment]
> That a UID exists in the node object, but the user dataset doesn't, is a failure on its own. This normally shouldn't happen. My concern is that if we catch it at this point, we'd obfuscate a data inconsistency - it might even get harder to discover the base problem.

I agree with this statement. There is a reason Backdrop gives you the following options when you try to delete a user:
- Delete the account and make its content belong to the Anonymous user
- Delete the account and its content

You can't have content that points at a user that doesn't exist in the site. You either have to make that content belong to anonymous, or delete that content. So, the steps to reproduce this issue (manually deleting a user from the table) should never happen in a normal operation of the site.

My suggestion is to revise/rethink the way you are importing your content. Don't import the UIDs created in a different site, which will obviously lead to huge problems, whether or not those uids exist in the new site. 

Doesn't Feeds allow you to specify the uid of imported content? Otherwise, just modify your data in Excel and change the uid of the content to be imported to point at a dummy user or anonymous (as you did).

**Comment by NormPlum** [comment]
> So, the steps to reproduce this issue (manually deleting a user from the table) should never happen in a normal operation of the site.

Yes. This issue is kind of like deleting the `system` table and then saying you're getting a bunch of errors. It is to be expected.
